### PR TITLE
[JSON Logic] Part 6: Inline rule handling for computedAttrs

### PR DESCRIPTION
**Pull request by brennj**
Adds the code needed to actually compute inline rules in `x-jsf-logic-computedAttrs` when they're defined in that keyword as opposed to in `x-jsf-logic`. Adds a bunch of tests and matching schemas to illustrate its capabilities.

### CODE REVIEW

File: src/tests/jsonLogicFixtures.js [hunk 1]
```diff
@@ -453,3 +435,132 @@ export const schemaWithPropertyThatDoesNotExistInThatLevelButDoesInFieldset = {
...
+export const schemaWithInlineMultipleRulesForComputedAttributes = {
+  properties: {
+    field_a: {
+      type: 'number',
+    },
+    field_b: {
+      type: 'number',
+      'x-jsf-logic-computedAttrs': {
+        description: {
+          value: 'Must be between {{half_a}} and {{double_a}}.',
```
Comment (sandrina-p) [inline src/tests/jsonLogicFixtures.js:485]: interpolation is one of my favorite features!! 🔥 

Note: I thought this would solve this [internal issue](https://linear.app/remote/issue/RMT-53), but although it's related it's not exactly the same.

File: src/tests/jsonLogicFixtures.js [hunk 2]
```diff
@@ -453,3 +435,132 @@ export const schemaWithPropertyThatDoesNotExistInThatLevelButDoesInFieldset = {
...
+          double_a: {
+            '*': [{ var: 'field_a' }, 2],
+          },
+        },
+      },
+    },
+  },
+};
+
+export const schemaSelfContainedValueForTitleWithNoTemplate = {
```
Comment (sandrina-p) [inline src/tests/jsonLogicFixtures.js:498]: question: What do you mean "withNoTemplate? It still uses `{{rule}}`.

File: src/tests/jsonLogicFixtures.js [hunk 3]
```diff
@@ -453,3 +435,132 @@ export const schemaWithPropertyThatDoesNotExistInThatLevelButDoesInFieldset = {
...
+  properties: {
+    field_a: {
+      type: 'number',
+    },
+    field_b: {
+      type: 'number',
+      'x-jsf-logic-computedAttrs': {
+        title: {
```
Comment (sandrina-p) [inline src/tests/jsonLogicFixtures.js:506]: thought (future iteration, not for now9: WDYT of a sugar syntax for this:

```
'x-jsf-logic-computedAttrs': {
  title: { '+': [{ var: 'field_a' }, 10] }
}
```

This can also lead us to computed min/max:

```
'x-jsf-logic-computedAttrs': {
  minimum: { '+': [{ var: 'field_a' }, 10] }
}
```

Which is an alternative to `x-jsf-logic-validations: ["bigger_than_a_plus_ten"]`

edit: I saw now the next example `schemaSelfContainedValueForMaximumMinimumValues`. Indeed it's an interesting case where we can add dynamic minimums in two ways. Either with `validations` or with `computations`. It's worth adding these to our guidelines — when to use one or another.

File: src/tests/jsonLogicFixtures.js [hunk 4]
```diff
@@ -453,3 +435,132 @@ export const schemaWithPropertyThatDoesNotExistInThatLevelButDoesInFieldset = {
...
+export const schemaWithJSFLogicAndInlineRule = {
+  properties: {
+    field_a: {
+      type: 'number',
+    },
+    field_b: {
+      type: 'number',
+      'x-jsf-logic-computedAttrs': {
+        title: {
+          value: 'Going to use {{rule}} and {{not_inline}}',
```
Comment (sandrina-p) [inline src/tests/jsonLogicFixtures.js:549]: praise: I can't express how powerful this is 🔥

File: src/tests/jsonLogic.test.js [hunk 5]
```diff
@@ -376,5 +381,69 @@ describe('cross-value validations', () => {
       expect(fieldB.maximum).toEqual(8);
       expect(fieldB.statement).toEqual({ description: 'Must be bigger than 4 and smaller than 8' });
     });
+
+    it('Use a self contained rule in a schema for a title attribute', () => {
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/tests/jsonLogic.test.js]: thats been updated, please let me know if you spot any other inconsistent use

File: src/tests/jsonLogicFixtures.js (hunk 4 above)
Comment (brennj) [reply src/tests/jsonLogicFixtures.js]: ❤️

File: src/tests/jsonLogic.test.js [hunk 6]
```diff
@@ -376,5 +381,69 @@ describe('cross-value validations', () => {
...
+        schemaSelfContainedValueForMaximumMinimumValues,
+        {
+          strictInputType: false,
+        }
+      );
+      expect(handleValidation({ field_a: 10, field_b: null }).formErrors).toEqual(undefined);
+      expect(handleValidation({ field_a: 50, field_b: 20 }).formErrors).toEqual({
+        field_b: 'Must be greater or equal to 40',
+      });
+      expect(handleValidation({ field_a: 50, field_b: 70 }).formErrors).toEqual({
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/tests/jsonLogic.test.js]: Is it? I'm getting the correct attributes on the field. Maybe it got fixed when I was doing other feedback in earlier PRs?

BTW, for the minimum NOT to be `NaN`, we need to add a default... i guess that makes sense as a tradeoff but i'll add a default: 0 to this example. Maybe we need to encourage this. When we merge this part into the docs, I'll add a note in the example internally.

File: src/tests/jsonLogicFixtures.js [hunk 7]
```diff
@@ -453,3 +435,132 @@ export const schemaWithPropertyThatDoesNotExistInThatLevelButDoesInFieldset = {
...
+          rule: {
+            '+': [{ var: 'field_a' }, 10],
+          },
+        },
+      },
+    },
+  },
+};
+
+export const schemaWithInlineRuleForComputedAttributeWithoutCopy = {
```
[the remark this answers is not in the document]
Comment (brennj) [reply src/tests/jsonLogicFixtures.js]: removing but i'm sure this will crop up in a later PR

**Comment by johnstonbl01** [comment]
This one looks good to me. Will defer to Sandrina on her previous comments.

File: src/tests/jsonLogic.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/tests/jsonLogic.test.js]: Yes, it's still missing the variables, for example schemaSelfContainedValueForTitle -> schemaInlineComputedAttrForTitle

File: src/tests/jsonLogicFixtures.js (hunk 3 above)
Comment (sandrina-p) [reply src/tests/jsonLogicFixtures.js]: No need for a sugar-syntax solution, this for MVP works perfectly. Regarding docs, do you already have a MR in place or just raw notes yet?

File: src/tests/jsonLogic.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/tests/jsonLogic.test.js]: Now it works on my side too! Pff perfect then!

[the remark this answers is not in the document]
Comment (sandrina-p) [reply src/tests/jsonLogic.test.js]: Regarding the "required" `default: 0` hummmm I don't like it. we are enforcing an optional property to makeup for a bug in json-logic 🙃 

Would [this my suggestion about nulls](https://github.com/remoteoss/json-schema-form/pull/41#discussion_r1311495011) solve the issue too? I need to leave in 5min, but I can test it myself tomorrow and try to come-up with another solution.

File: src/tests/jsonLogic.test.js [hunk 8]
```diff
@@ -344,5 +349,72 @@ describe('jsonLogic: cross-values validations', () => {
       expect(fieldB.maximum).toEqual(8);
       expect(fieldB.statement).toEqual({ description: 'Must be bigger than 4 and smaller than 8' });
     });
+
```
Comment (sandrina-p) [inline src/tests/jsonLogic.test.js:352]: > [@johnstonbl01] This one looks good to me. Will defer to Sandrina on her previous comments.

We still are on Part 3, but please _never ever_ let my comments (and long OOO/sick/moving) be your bottleneck.

If you don't trust your gut to approve this, trust the automated tests. For real, the tests are one of the things I am most proud of about this library. They cover a lot, and we also have Integration tests (Storybook+React). They are what give me full confidence that this is working and won't break previous versions.

In other words: _trust the tests, not me._ ✌️

File: src/tests/jsonLogic.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (brennj) [reply src/tests/jsonLogic.test.js]: good catch, updated

File: src/tests/jsonLogicFixtures.js (hunk 3 above)
Comment (brennj) [reply src/tests/jsonLogicFixtures.js]: I'm thinking the RFC, minus the research parts are going to be pretty useful as a start. Lets think about that in a bit.

Comment (johnstonbl01) [reply src/tests/jsonLogicFixtures.js]: Agree that the RFC is a good starting point. 👍

**Review by johnstonbl01 — APPROVED** [review]

**Review by sandrina-p — APPROVED** [review]

**Review by dilvane — APPROVED** [review]
